python-spec/src/somacore/query/query.py

```diff
@@ -434,6 +456,47 @@ def _axisp_inner(
         joinids = getattr(self._joinids, axis.value)
         return axisp[layer].read((joinids, joinids))
 
+    def _axism_inner(
+        self,
+        axis: "_Axis",
+        layer: str,
+    ) -> data.SparseRead:
+        key = axis.value + "m"
+
+        if key not in self._ms:
+            raise ValueError(f"Measurement does not contain {key} data")
+
+        axism = self._ms.obsm if axis is _Axis.OBS else self._ms.varm
+        if not (layer and layer in axism):
+            raise ValueError(f"Must specify '{key}' layer")
+
+        n_row, n_col = axism[layer].shape
+        col_joinids = pa.array(range(n_col))
+
+        joinids = getattr(self._joinids, axis.value)
+        return axism[layer].read((joinids, col_joinids))
+
+    def _axism_inner_csr(
```

Comment:
This method exists purely to be passed to `_read`, which loads everything in memory. This works for now but I am not sure this is the right format we want for `obsm`. In principle, for `obsm` it would be better to use a csc matrix since it has way more rows than column. Most embeddings though are dense, so maybe we could just use a numpy array here?

Reply:
I suggest looking at the AnnData documentation.  obsm/varm/obsp/varp are ndarray in typical use, and I don't see any reason they would not be here...  Using a sparse matrix is likely to create incompatibility with downstream user assumptions.

Reply:
Good point, for some reason I thought AnnData didn't enforce the typing here, but it looks like it does. I'll convert to ndarray.

Reply:
its more than "enforce typing" - it is conceptually a dense array in the AnnData data model (but not in SOMA)